File: src/google_slides_mcp/tools/analysis.py
"""Presentation analysis tools for Google Slides.

Tools for deep-diving into presentations to extract style guides,
structural patterns, and usage recommendations.
"""

from typing import TYPE_CHECKING

from fastmcp import Context

if TYPE_CHECKING:
    from fastmcp import FastMCP


def register_analysis_tools(mcp: "FastMCP") -> None:
    """Register analysis tools with the MCP application.

    Args:
        mcp: The FastMCP application instance
    """

    @mcp.tool()
    async def analyze_presentation(
        ctx: Context,
        presentation_id: str,
        include_thumbnails: bool = False,
        max_thumbnail_slides: int = 5,
    ) -> dict:
        """Deep-dive analysis of a presentation to extract style guide information.

        This tool analyzes a presentation's structure, styling, colors, fonts,
        placeholders, and layouts to provide comprehensive guidance for
        programmatic presentation generation.

        Use this when you need to understand a template before using it, or
        to extract styling patterns for consistent presentation creation.

        Args:
            presentation_id: The presentation ID to analyze
            include_thumbnails: Whether to generate thumbnail URLs for key slides
            max_thumbnail_slides: Maximum number of thumbnails to generate (1-10)

        Returns:
            Dictionary with comprehensive analysis:
            - overview: Presentation metadata and statistics
            - slide_inventory: Categorized list of all slides
            - color_palette: Unique colors found with usage context
            - typography: Fonts, sizes, and text styles detected
            - placeholder_patterns: Common placeholder text patterns
            - layout_categories: Slides grouped by layout type
            - recommendations: Suggestions for programmatic usage
            - thumbnails: URLs to key slide thumbnails (if requested)
        """
        from google_slides_mcp.auth.middleware import GoogleAuthMiddleware
        from google_slides_mcp.services.slides_service import SlidesService
        from google_slides_mcp.utils.units import emu_to_inches

        middleware = GoogleAuthMiddleware()
        credentials = await middleware.extract_credentials(ctx)
        service = SlidesService(credentials)

        # Get full presentation data
        presentation = await service.get_presentation(presentation_id)

        # Initialize analysis containers
        colors_found: dict[str, list[str]] = {}  # color -> [contexts]
        fonts_found: dict[str, list[str]] = {}   # font -> [contexts]
        font_sizes: dict[float, int] = {}         # size -> count
        placeholder_texts: list[dict] = []
        slide_categories: dict[str, list[dict]] = {
            "cover": [],
            "section_divider": [],
            "content": [],
            "image_focused": [],
            "data_visualization": [],
            "mockup": [],
            "infographic": [],
            "other": [],
        }

        # Extract page size
        page_size = presentation.get("pageSize", {})
        width_emu = page_size.get("width", {}).get("magnitude", 0)
        height_emu = page_size.get("height", {}).get("magnitude", 0)
        width_inches = emu_to_inches(width_emu)
        height_inches = emu_to_inches(height_emu)

        # Determine aspect ratio
        if width_inches > 0 and height_inches > 0:
            ratio = width_inches / height_inches
            if abs(ratio - 16/9) < 0.1:
                aspect_ratio = "16:9 (Widescreen)"
            elif abs(ratio - 4/3) < 0.1:
                aspect_ratio = "4:3 (Standard)"
            elif abs(ratio - 16/10) < 0.1:
                aspect_ratio = "16:10"
            else:
                aspect_ratio = f"{ratio:.2f}:1 (Custom)"
        else:
            aspect_ratio = "Unknown"

        slides = presentation.get("slides", [])
        slides_info = []

        for i, slide in enumerate(slides):
            slide_id = slide.get("objectId", "")
            page_elements = slide.get("pageElements", [])
            element_count = len(page_elements)

            # Extract slide info
            slide_info = {
                "slide_id": slide_id,
                "index": i,
                "element_count": element_count,
                "title": "",
                "subtitle": "",
                "has_image": False,
                "has_chart": False,
                "has_table": False,
                "placeholder_types": [],
            }

            # Analyze each element
            for element in page_elements:
                # Check element types
                if "image" in element:
                    slide_info["has_image"] = True
                elif "sheetsChart" in element:
                    slide_info["has_chart"] = True
                elif "table" in element:
                    slide_info["has_table"] = True

                # Analyze shapes
                shape = element.get("shape", {})
                if shape:
                    placeholder = shape.get("placeholder", {})
                    placeholder_type = placeholder.get("type")

                    if placeholder_type:
                        slide_info["placeholder_types"].append(placeholder_type)

                    # Extract text content
                    text_data = shape.get("text", {})
                    text_elements = text_data.get("textElements", [])

                    for text_elem in text_elements:
                        text_run = text_elem.get("textRun", {})
                        content = text_run.get("content", "").strip()
                        style = text_run.get("style", {})

                        if content:
                            # Store placeholder text
                            if placeholder_type:
                                if placeholder_type == "TITLE":
                                    slide_info["title"] = content
                                elif placeholder_type == "SUBTITLE":
                                    slide_info["subtitle"] = content

                                placeholder_texts.append({
                                    "type": placeholder_type,
                                    "text": content[:100],  # Truncate
                                    "slide_index": i,
                                })

                            # Extract font info
                            font_family = style.get("fontFamily")
                            if font_family:
                                if font_family not in fonts_found:
                                    fonts_found[font_family] = []
                                context = f"Slide {i+1}"
                                if context not in fonts_found[font_family]:
                                    fonts_found[font_family].append(context)

                            # Extract font size
                            font_size = style.get("fontSize", {}).get("magnitude")
                            if font_size:
                                font_sizes[font_size] = font_sizes.get(font_size, 0) + 1

                            # Extract text color
                            fg_color = style.get("foregroundColor", {}).get("opaqueColor", {})
                            _extract_color(fg_color, f"Text on slide {i+1}", colors_found)

                    # Extract shape colors
                    shape_props = shape.get("shapeProperties", {})
                    bg_fill = shape_props.get("shapeBackgroundFill", {})
                    solid_fill = bg_fill.get("solidFill", {})
                    if solid_fill:
                        color_data = solid_fill.get("color", {})
                        _extract_color(color_data, f"Shape fill on slide {i+1}", colors_found)

                # Extract background colors from page properties
                page_props = slide.get("pageProperties", {})
                bg_fill = page_props.get("pageBackgroundFill", {})
                solid_fill = bg_fill.get("solidFill", {})
                if solid_fill:
                    color_data = solid_fill.get("color", {})
                    _extract_color(color_data, f"Background on slide {i+1}", colors_found)

            # Categorize slide
            category = _categorize_slide(slide_info)
            slide_info["category"] = category
            slide_categories[category].append({
                "index": i,
                "slide_id": slide_id,
                "title": slide_info["title"],
            })

            slides_info.append(slide_info)

        # Generate placeholder pattern analysis
        placeholder_patterns = _analyze_placeholder_patterns(placeholder_texts)

        # Build recommendations
        recommendations = _generate_recommendations(
            slides_info, placeholder_patterns, fonts_found, colors_found
        )

        # Generate thumbnails if requested
        thumbnails = []
        if include_thumbnails:
            from googleapiclient.discovery import build
            slides_service = build("slides", "v1", credentials=credentials)

            # Select key slides for thumbnails
            key_slides = _select_key_slides(slides_info, max_thumbnail_slides)
            for slide_info in key_slides:
                try:
                    thumbnail = (
                        slides_service.presentations()
                        .pages()
                        .getThumbnail(
                            presentationId=presentation_id,
                            pageObjectId=slide_info["slide_id"],
                            thumbnailProperties_mimeType="PNG",
                        )
                        .execute()
                    )
                    thumbnails.append({
                        "slide_index": slide_info["index"],
                        "slide_id": slide_info["slide_id"],
                        "title": slide_info["title"],
                        "category": slide_info.get("category", "unknown"),
                        "url": thumbnail.get("contentUrl", ""),
                    })
                except Exception:
                    pass  # Skip failed thumbnails

        # Build color palette summary
        color_palette = []
        for color, contexts in colors_found.items():
            color_palette.append({
                "color": color,
                "usage_count": len(contexts),
                "contexts": contexts[:5],  # Limit context examples
            })
        color_palette.sort(key=lambda x: x["usage_count"], reverse=True)

        # Build typography summary
        typography = {
            "fonts": list(fonts_found.keys()),
            "primary_font": max(fonts_found.keys(), key=lambda f: len(fonts_found[f])) if fonts_found else None,
            "font_sizes": [
                {"size_pt": size, "count": count}
                for size, count in sorted(font_sizes.items(), key=lambda x: -x[1])[:10]
            ],
        }

        # Build layout categories summary
        layout_summary = {}
        for category, slides_list in slide_categories.items():
            if slides_list:
                layout_summary[category] = {
                    "count": len(slides_list),
                    "slides": slides_list[:10],  # Limit examples
                }

        return {
            "overview": {
                "presentation_id": presentation_id,
                "title": presentation.get("title", "Untitled"),
                "total_slides": len(slides),
                "page_size": {
                    "width_inches": round(width_inches, 2),
                    "height_inches": round(height_inches, 2),
                },
                "aspect_ratio": aspect_ratio,
                "url": f"https://docs.google.com/presentation/d/{presentation_id}",
            },
            "slide_inventory": slides_info,
            "color_palette": color_palette[:20],  # Top 20 colors
            "typography": typography,
            "placeholder_patterns": placeholder_patterns,
            "layout_categories": layout_summary,
            "recommendations": recommendations,
            "thumbnails": thumbnails if include_thumbnails else None,
        }


def _extract_color(color_data: dict, context: str, colors_found: dict) -> None:
    """Extract color from Google Slides color data and add to colors_found."""
    rgb = color_data.get("rgbColor", {})
    theme_color = color_data.get("themeColor")

    if theme_color:
        color_key = f"theme:{theme_color}"
    elif rgb:
        r = int(rgb.get("red", 0) * 255)
        g = int(rgb.get("green", 0) * 255)
        b = int(rgb.get("blue", 0) * 255)
        color_key = f"#{r:02x}{g:02x}{b:02x}"
    else:
        return

    if color_key not in colors_found:
        colors_found[color_key] = []
    if context not in colors_found[color_key]:
        colors_found[color_key].append(context)


def _categorize_slide(slide_info: dict) -> str:
    """Categorize a slide based on its content and structure."""
    title = slide_info["title"].lower()
    placeholders = slide_info["placeholder_types"]
    element_count = slide_info["element_count"]

    # Cover detection
    if any(kw in title for kw in ["cover", "title page"]) or (
        element_count <= 4 and "TITLE" in placeholders and "BODY" in placeholders
        and slide_info["index"] < 15
    ):
        return "cover"

    # Section divider detection
    if any(kw in title for kw in ["section", "divider"]) or (
        element_count <= 3 and "TITLE" in placeholders
        and ("SUBTITLE" in placeholders or element_count == 1)
    ):
        return "section_divider"

    # Data visualization
    if slide_info["has_chart"] or any(kw in title for kw in ["chart", "graph", "table", "data"]):
        return "data_visualization"

    if slide_info["has_table"]:
        return "data_visualization"

    # Infographic detection
    if "infographic" in title or element_count > 15:
        return "infographic"

    # Mockup detection
    if any(kw in title for kw in ["mockup", "phone", "laptop", "device", "smartphone", "notebook"]):
        return "mockup"

    # Image-focused
    if slide_info["has_image"] and element_count <= 5:
        return "image_focused"

    # Content (default for slides with substantial content)
    if "BODY" in placeholders or element_count >= 3:
        return "content"

    return "other"


def _analyze_placeholder_patterns(placeholder_texts: list[dict]) -> dict:
    """Analyze placeholder texts to find common patterns."""
    patterns = {
        "title_patterns": [],
        "subtitle_patterns": [],
        "body_patterns": [],
        "date_patterns": [],
        "name_patterns": [],
    }

    for item in placeholder_texts:
        text = item["text"]
        ptype = item["type"]

        # Detect date patterns
        if any(dp in text for dp in ["MM.DD", "YYYY", "mm/dd", "date"]):
            if text not in [p["text"] for p in patterns["date_patterns"]]:
                patterns["date_patterns"].append({"text": text, "type": ptype})

        # Detect name patterns
        if any(np in text.lower() for np in ["full name", "name //", "job title"]):
            if text not in [p["text"] for p in patterns["name_patterns"]]:
                patterns["name_patterns"].append({"text": text, "type": ptype})

        # Collect by placeholder type
        if ptype == "TITLE":
            if len(patterns["title_patterns"]) < 10:
                patterns["title_patterns"].append(text)
        elif ptype == "SUBTITLE":
            if len(patterns["subtitle_patterns"]) < 10:
                patterns["subtitle_patterns"].append(text)
        elif ptype == "BODY":
            # Only store first 50 chars of body
            if len(patterns["body_patterns"]) < 5:
                patterns["body_patterns"].append(text[:50] + "..." if len(text) > 50 else text)

    return patterns


def _generate_recommendations(
    slides_info: list[dict],
    placeholder_patterns: dict,
    fonts_found: dict,
    colors_found: dict,
) -> list[str]:
    """Generate usage recommendations based on analysis."""
    recommendations = []

    # Slide usage recommendations
    cover_slides = [s for s in slides_info if s.get("category") == "cover"]
    if cover_slides:
        recommendations.append(
            f"Use slides {', '.join(str(s['index']+1) for s in cover_slides[:4])} "
            f"as cover options (found {len(cover_slides)} cover variants)"
        )

    section_slides = [s for s in slides_info if s.get("category") == "section_divider"]
    if section_slides:
        recommendations.append(
            f"Use slides {', '.join(str(s['index']+1) for s in section_slides[:2])} "
            f"as section dividers"
        )

    # Placeholder recommendations
    if placeholder_patterns.get("date_patterns"):
        date_example = placeholder_patterns["date_patterns"][0]["text"]
        recommendations.append(
            f"Replace date placeholder '{date_example}' with actual dates"
        )

    if placeholder_patterns.get("name_patterns"):
        name_example = placeholder_patterns["name_patterns"][0]["text"]
        recommendations.append(
            f"Replace name placeholder '{name_example}' with presenter info"
        )

    # Font recommendation
    if fonts_found:
        primary_font = max(fonts_found.keys(), key=lambda f: len(fonts_found[f]))
        recommendations.append(
            f"Maintain '{primary_font}' as the primary font for consistency"
        )

    # Color recommendation
    if colors_found:
        top_colors = sorted(colors_found.keys(), key=lambda c: len(colors_found[c]), reverse=True)[:3]
        recommendations.append(
            f"Primary brand colors detected: {', '.join(top_colors)}"
        )

    # Workflow recommendation
    recommendations.append(
        "Workflow: copy_template → delete unused slides → replace_placeholders → add images"
    )

    return recommendations


def _select_key_slides(slides_info: list[dict], max_count: int) -> list[dict]:
    """Select key representative slides for thumbnails."""
    key_slides = []
    categories_seen = set()

    # Prioritize one from each category
    priority_order = ["cover", "section_divider", "content", "data_visualization", "mockup", "infographic"]

    for category in priority_order:
        if len(key_slides) >= max_count:
            break
        for slide in slides_info:
            if slide.get("category") == category and category not in categories_seen:
                key_slides.append(slide)
                categories_seen.add(category)
                break

    # Fill remaining slots with other slides
    for slide in slides_info:
        if len(key_slides) >= max_count:
            break
        if slide not in key_slides:
            key_slides.append(slide)

    return key_slides[:max_count]

File: packages/cloudflare/src/tools/analysis.ts
/**
 * Presentation analysis tools for Google Slides.
 *
 * Tools for deep-diving into presentations to extract style guides,
 * structural patterns, and usage recommendations.
 */

import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import { SlidesClient } from "../api/slides-client.js";
import type { Props } from "../types.js";
import type { Page, Presentation } from "../api/types.js";
import { emuToInches } from "../utils/units.js";

interface SlideInfo {
  slide_id: string;
  index: number;
  element_count: number;
  title: string;
  subtitle: string;
  has_image: boolean;
  has_chart: boolean;
  has_table: boolean;
  placeholder_types: string[];
  category?: string;
}

interface PlaceholderText {
  type: string;
  text: string;
  slide_index: number;
}

/**
 * Extract color from Google Slides color data.
 */
function extractColor(
  colorData: Record<string, unknown>,
  context: string,
  colorsFound: Map<string, string[]>
): void {
  const rgb = colorData.rgbColor as Record<string, number> | undefined;
  const themeColor = colorData.themeColor as string | undefined;

  let colorKey: string;
  if (themeColor) {
    colorKey = `theme:${themeColor}`;
  } else if (rgb) {
    const r = Math.floor((rgb.red ?? 0) * 255);
    const g = Math.floor((rgb.green ?? 0) * 255);
    const b = Math.floor((rgb.blue ?? 0) * 255);
    colorKey = `#${r.toString(16).padStart(2, "0")}${g.toString(16).padStart(2, "0")}${b.toString(16).padStart(2, "0")}`;
  } else {
    return;
  }

  const contexts = colorsFound.get(colorKey) ?? [];
  if (!contexts.includes(context)) {
    contexts.push(context);
    colorsFound.set(colorKey, contexts);
  }
}

/**
 * Categorize a slide based on its content and structure.
 */
function categorizeSlide(slideInfo: SlideInfo): string {
  const title = slideInfo.title.toLowerCase();
  const placeholders = slideInfo.placeholder_types;
  const elementCount = slideInfo.element_count;

  // Cover detection
  if (
    title.includes("cover") ||
    title.includes("title page") ||
    (elementCount <= 4 &&
      placeholders.includes("TITLE") &&
      placeholders.includes("BODY") &&
      slideInfo.index < 15)
  ) {
    return "cover";
  }

  // Section divider detection
  if (
    title.includes("section") ||
    title.includes("divider") ||
    (elementCount <= 3 &&
      placeholders.includes("TITLE") &&
      (placeholders.includes("SUBTITLE") || elementCount === 1))
  ) {
    return "section_divider";
  }

  // Data visualization
  if (
    slideInfo.has_chart ||
    title.includes("chart") ||
    title.includes("graph") ||
    title.includes("table") ||
    title.includes("data")
  ) {
    return "data_visualization";
  }

  if (slideInfo.has_table) {
    return "data_visualization";
  }

  // Infographic detection
  if (title.includes("infographic") || elementCount > 15) {
    return "infographic";
  }

  // Mockup detection
  if (
    title.includes("mockup") ||
    title.includes("phone") ||
    title.includes("laptop") ||
    title.includes("device")
  ) {
    return "mockup";
  }

  // Image-focused
  if (slideInfo.has_image && elementCount <= 5) {
    return "image_focused";
  }

  // Content (default for slides with substantial content)
  if (placeholders.includes("BODY") || elementCount >= 3) {
    return "content";
  }

  return "other";
}

/**
 * Analyze placeholder texts to find common patterns.
 */
function analyzePlaceholderPatterns(placeholderTexts: PlaceholderText[]): Record<string, unknown> {
  const patterns: Record<string, unknown[]> = {
    title_patterns: [],
    subtitle_patterns: [],
    body_patterns: [],
    date_patterns: [],
    name_patterns: [],
  };

  for (const item of placeholderTexts) {
    const text = item.text;
    const ptype = item.type;

    // Detect date patterns
    if (
      text.includes("MM.DD") ||
      text.includes("YYYY") ||
      text.includes("mm/dd") ||
      text.toLowerCase().includes("date")
    ) {
      const datePatterns = patterns.date_patterns as Array<{ text: string; type: string }>;
      if (!datePatterns.some((p) => p.text === text)) {
        datePatterns.push({ text, type: ptype });
      }
    }

    // Detect name patterns
    if (
      text.toLowerCase().includes("full name") ||
      text.includes("name //") ||
      text.toLowerCase().includes("job title")
    ) {
      const namePatterns = patterns.name_patterns as Array<{ text: string; type: string }>;
      if (!namePatterns.some((p) => p.text === text)) {
        namePatterns.push({ text, type: ptype });
      }
    }

    // Collect by placeholder type
    if (ptype === "TITLE") {
      const titlePatterns = patterns.title_patterns as string[];
      if (titlePatterns.length < 10) {
        titlePatterns.push(text);
      }
    } else if (ptype === "SUBTITLE") {
      const subtitlePatterns = patterns.subtitle_patterns as string[];
      if (subtitlePatterns.length < 10) {
        subtitlePatterns.push(text);
      }
    } else if (ptype === "BODY") {
      const bodyPatterns = patterns.body_patterns as string[];
      if (bodyPatterns.length < 5) {
        bodyPatterns.push(text.length > 50 ? text.substring(0, 50) + "..." : text);
      }
    }
  }

  return patterns;
}

/**
 * Generate usage recommendations based on analysis.
 */
function generateRecommendations(
  slidesInfo: SlideInfo[],
  placeholderPatterns: Record<string, unknown>,
  fontsFound: Map<string, string[]>,
  colorsFound: Map<string, string[]>
): string[] {
  const recommendations: string[] = [];

  // Slide usage recommendations
  const coverSlides = slidesInfo.filter((s) => s.category === "cover");
  if (coverSlides.length > 0) {
    const indices = coverSlides.slice(0, 4).map((s) => s.index + 1).join(", ");
    recommendations.push(
      `Use slides ${indices} as cover options (found ${coverSlides.length} cover variants)`
    );
  }

  const sectionSlides = slidesInfo.filter((s) => s.category === "section_divider");
  if (sectionSlides.length > 0) {
    const indices = sectionSlides.slice(0, 2).map((s) => s.index + 1).join(", ");
    recommendations.push(`Use slides ${indices} as section dividers`);
  }

  // Placeholder recommendations
  const datePatterns = placeholderPatterns.date_patterns as Array<{ text: string }>;
  if (datePatterns && datePatterns.length > 0) {
    recommendations.push(
      `Replace date placeholder '${datePatterns[0].text}' with actual dates`
    );
  }

  const namePatterns = placeholderPatterns.name_patterns as Array<{ text: string }>;
  if (namePatterns && namePatterns.length > 0) {
    recommendations.push(
      `Replace name placeholder '${namePatterns[0].text}' with presenter info`
    );
  }

  // Font recommendation
  if (fontsFound.size > 0) {
    let primaryFont = "";
    let maxUsage = 0;
    for (const [font, contexts] of fontsFound) {
      if (contexts.length > maxUsage) {
        maxUsage = contexts.length;
        primaryFont = font;
      }
    }
    if (primaryFont) {
      recommendations.push(
        `Maintain '${primaryFont}' as the primary font for consistency`
      );
    }
  }

  // Color recommendation
  if (colorsFound.size > 0) {
    const sortedColors = [...colorsFound.entries()]
      .sort((a, b) => b[1].length - a[1].length)
      .slice(0, 3)
      .map(([color]) => color);
    recommendations.push(`Primary brand colors detected: ${sortedColors.join(", ")}`);
  }

  // Workflow recommendation
  recommendations.push(
    "Workflow: copy_template → delete unused slides → replace_placeholders → add images"
  );

  return recommendations;
}

/**
 * Select key representative slides for thumbnails.
 */
function selectKeySlides(slidesInfo: SlideInfo[], maxCount: number): SlideInfo[] {
  const keySlides: SlideInfo[] = [];
  const categoriesSeen = new Set<string>();

  const priorityOrder = [
    "cover",
    "section_divider",
    "content",
    "data_visualization",
    "mockup",
    "infographic",
  ];

  for (const category of priorityOrder) {
    if (keySlides.length >= maxCount) break;
    for (const slide of slidesInfo) {
      if (slide.category === category && !categoriesSeen.has(category)) {
        keySlides.push(slide);
        categoriesSeen.add(category);
        break;
      }
    }
  }

  // Fill remaining slots
  for (const slide of slidesInfo) {
    if (keySlides.length >= maxCount) break;
    if (!keySlides.includes(slide)) {
      keySlides.push(slide);
    }
  }

  return keySlides.slice(0, maxCount);
}

/**
 * Register analysis tools with the MCP server.
 */
export function registerAnalysisTools(
  server: McpServer,
  _env: Env,
  props: Props
): void {
  const client = new SlidesClient({ accessToken: props.accessToken });

  /**
   * analyze_presentation - Deep-dive analysis of a presentation
   */
  server.tool(
    "analyze_presentation",
    "Deep-dive analysis of a presentation to extract style guide information. Analyzes structure, styling, colors, fonts, placeholders, and layouts.",
    {
      presentation_id: z.string().describe("The presentation ID to analyze"),
      include_thumbnails: z.boolean().default(false).describe("Whether to generate thumbnail URLs for key slides"),
      max_thumbnail_slides: z.number().min(1).max(10).default(5).describe("Maximum number of thumbnails to generate"),
    },
    async ({ presentation_id, include_thumbnails, max_thumbnail_slides }) => {
      try {
        const presentation = await client.getPresentation(presentation_id);

        // Initialize analysis containers
        const colorsFound = new Map<string, string[]>();
        const fontsFound = new Map<string, string[]>();
        const fontSizes = new Map<number, number>();
        const placeholderTexts: PlaceholderText[] = [];
        const slideCategories: Record<string, Array<{ index: number; slide_id: string; title: string }>> = {
          cover: [],
          section_divider: [],
          content: [],
          image_focused: [],
          data_visualization: [],
          mockup: [],
          infographic: [],
          other: [],
        };

        // Extract page size
        const pageSize = presentation.pageSize;
        const widthEmu = pageSize?.width?.magnitude ?? 0;
        const heightEmu = pageSize?.height?.magnitude ?? 0;
        const widthInches = emuToInches(widthEmu);
        const heightInches = emuToInches(heightEmu);

        // Determine aspect ratio
        let aspectRatio: string;
        if (widthInches > 0 && heightInches > 0) {
          const ratio = widthInches / heightInches;
          if (Math.abs(ratio - 16 / 9) < 0.1) {
            aspectRatio = "16:9 (Widescreen)";
          } else if (Math.abs(ratio - 4 / 3) < 0.1) {
            aspectRatio = "4:3 (Standard)";
          } else if (Math.abs(ratio - 16 / 10) < 0.1) {
            aspectRatio = "16:10";
          } else {
            aspectRatio = `${ratio.toFixed(2)}:1 (Custom)`;
          }
        } else {
          aspectRatio = "Unknown";
        }

        const slides = presentation.slides ?? [];
        const slidesInfo: SlideInfo[] = [];

        for (let i = 0; i < slides.length; i++) {
          const slide = slides[i];
          const slideId = slide.objectId ?? "";
          const pageElements = slide.pageElements ?? [];
          const elementCount = pageElements.length;

          const slideInfo: SlideInfo = {
            slide_id: slideId,
            index: i,
            element_count: elementCount,
            title: "",
            subtitle: "",
            has_image: false,
            has_chart: false,
            has_table: false,
            placeholder_types: [],
          };

          // Analyze each element
          for (const element of pageElements) {
            // Check element types
            if (element.image) {
              slideInfo.has_image = true;
            } else if ((element as unknown as Record<string, unknown>).sheetsChart) {
              slideInfo.has_chart = true;
            } else if (element.table) {
              slideInfo.has_table = true;
            }

            // Analyze shapes
            const shape = element.shape;
            if (shape) {
              const placeholder = shape.placeholder;
              const placeholderType = placeholder?.type;

              if (placeholderType) {
                slideInfo.placeholder_types.push(placeholderType);
              }

              // Extract text content
              const textElements = shape.text?.textElements ?? [];

              for (const textElem of textElements) {
                const textRun = textElem.textRun;
                const content = (textRun?.content ?? "").trim();
                const style = textRun?.style;

                if (content) {
                  // Store placeholder text
                  if (placeholderType) {
                    if (placeholderType === "TITLE") {
                      slideInfo.title = content;
                    } else if (placeholderType === "SUBTITLE") {
                      slideInfo.subtitle = content;
                    }

                    placeholderTexts.push({
                      type: placeholderType,
                      text: content.substring(0, 100),
                      slide_index: i,
                    });
                  }

                  // Extract font info
                  const fontFamily = style?.fontFamily;
                  if (fontFamily) {
                    const contexts = fontsFound.get(fontFamily) ?? [];
                    const context = `Slide ${i + 1}`;
                    if (!contexts.includes(context)) {
                      contexts.push(context);
                      fontsFound.set(fontFamily, contexts);
                    }
                  }

                  // Extract font size
                  const fontSize = style?.fontSize?.magnitude;
                  if (fontSize) {
                    fontSizes.set(fontSize, (fontSizes.get(fontSize) ?? 0) + 1);
                  }

                  // Extract text color
                  const fgColor = style?.foregroundColor?.opaqueColor;
                  if (fgColor) {
                    extractColor(
                      fgColor as unknown as Record<string, unknown>,
                      `Text on slide ${i + 1}`,
                      colorsFound
                    );
                  }
                }
              }

              // Extract shape colors
              const shapeProps = shape.shapeProperties;
              const bgFill = shapeProps?.shapeBackgroundFill;
              const solidFill = (bgFill as Record<string, unknown>)?.solidFill as Record<string, unknown> | undefined;
              if (solidFill?.color) {
                extractColor(
                  solidFill.color as Record<string, unknown>,
                  `Shape fill on slide ${i + 1}`,
                  colorsFound
                );
              }
            }
          }

          // Categorize slide
          const category = categorizeSlide(slideInfo);
          slideInfo.category = category;
          slideCategories[category].push({
            index: i,
            slide_id: slideId,
            title: slideInfo.title,
          });

          slidesInfo.push(slideInfo);
        }

        // Generate placeholder pattern analysis
        const placeholderPatterns = analyzePlaceholderPatterns(placeholderTexts);

        // Build recommendations
        const recommendations = generateRecommendations(
          slidesInfo,
          placeholderPatterns,
          fontsFound,
          colorsFound
        );

        // Generate thumbnails if requested
        const thumbnails: Array<{
          slide_index: number;
          slide_id: string;
          title: string;
          category: string;
          url: string;
        }> = [];

        if (include_thumbnails) {
          const keySlides = selectKeySlides(slidesInfo, max_thumbnail_slides);
          for (const slideInfo of keySlides) {
            try {
              const thumbnail = await client.getThumbnail(
                presentation_id,
                slideInfo.slide_id,
                "PNG"
              );
              thumbnails.push({
                slide_index: slideInfo.index,
                slide_id: slideInfo.slide_id,
                title: slideInfo.title,
                category: slideInfo.category ?? "unknown",
                url: thumbnail.contentUrl ?? "",
              });
            } catch {
              // Skip failed thumbnails
            }
          }
        }

        // Build color palette summary
        const colorPalette = [...colorsFound.entries()]
          .map(([color, contexts]) => ({
            color,
            usage_count: contexts.length,
            contexts: contexts.slice(0, 5),
          }))
          .sort((a, b) => b.usage_count - a.usage_count)
          .slice(0, 20);

        // Build typography summary
        let primaryFont: string | null = null;
        let maxFontUsage = 0;
        for (const [font, contexts] of fontsFound) {
          if (contexts.length > maxFontUsage) {
            maxFontUsage = contexts.length;
            primaryFont = font;
          }
        }

        const typography = {
          fonts: [...fontsFound.keys()],
          primary_font: primaryFont,
          font_sizes: [...fontSizes.entries()]
            .sort((a, b) => b[1] - a[1])
            .slice(0, 10)
            .map(([size, count]) => ({ size_pt: size, count })),
        };

        // Build layout categories summary
        const layoutSummary: Record<string, { count: number; slides: unknown[] }> = {};
        for (const [category, slidesList] of Object.entries(slideCategories)) {
          if (slidesList.length > 0) {
            layoutSummary[category] = {
              count: slidesList.length,
              slides: slidesList.slice(0, 10),
            };
          }
        }

        return {
          content: [
            {
              type: "text" as const,
              text: JSON.stringify(
                {
                  overview: {
                    presentation_id,
                    title: presentation.title ?? "Untitled",
                    total_slides: slides.length,
                    page_size: {
                      width_inches: Math.round(widthInches * 100) / 100,
                      height_inches: Math.round(heightInches * 100) / 100,
                    },
                    aspect_ratio: aspectRatio,
                    url: `https://docs.google.com/presentation/d/${presentation_id}`,
                  },
                  slide_inventory: slidesInfo,
                  color_palette: colorPalette,
                  typography,
                  placeholder_patterns: placeholderPatterns,
                  layout_categories: layoutSummary,
                  recommendations,
                  thumbnails: include_thumbnails ? thumbnails : null,
                },
                null,
                2
              ),
            },
          ],
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text" as const,
              text: `Error: ${error instanceof Error ? error.message : String(error)}`,
            },
          ],
          isError: true,
        };
      }
    }
  );
}
